### ISSUE TITLE
Stateful model with keras section returns loss: nan

### ISSUE THREAD

**Issue by dov** [roebius/deeplearning1_keras2#7]
I'm using the theano backend, and the LSTM example does not work. The statements:

```
model=Sequential([
        Embedding(vocab_size, n_fac, input_length=cs, batch_input_shape=(bs,8)),
        BatchNormalization(),
        LSTM(n_hidden, return_sequences=True, stateful=True),
        TimeDistributed(Dense(vocab_size, activation='softmax')),
    ])
model.compile(loss='sparse_categorical_crossentropy', optimizer=Adam())
mx = len(x_rnn)//bs*bs
model.fit(x_rnn[:mx], y_rnn[:mx], batch_size=bs, epochs=4, shuffle=False)
```

returns outputwith a nan-value for the loss.
```
Epoch 1/4
75072/75072 [==============================] - 20s - loss: nan     
Epoch 2/4
75072/75072 [==============================] - 20s - loss: nan     
Epoch 3/4
75072/75072 [==============================] - 20s - loss: nan     
Epoch 4/4
75072/75072 [==============================] - 20s - loss: nan
```

**Comment by roebius** [comment]
At first glance it could be an issue due to a too high learning rate. I believe Adam's default in Keras is 1e-3. Could you make a test reducing the learning rate to 1e-4 or less just before calling model.fit?
_Example_
`K.set_value(model.optimizer.lr, 1e-4)`
`model.fit(x_rnn[:mx], y_rnn[:mx], batch_size=bs, epochs=4, shuffle=False)`

If this doesn't work, checking your version of both Theano and Keras would help me further investigate.

**Comment by dov** [comment]
I tried to make a minimal working example to reproduce the problem:

```
import keras
from keras.models import Sequential, Model
from keras.layers import Embedding, LSTM, TimeDistributed, Dense
from keras.layers.normalization import BatchNormalization
from keras.layers.core import Dense
from keras.optimizers import Adam
import numpy as np

path = keras.utils.data_utils.get_file('nietzsche.txt',
    origin="https://s3.amazonaws.com/text-datasets/nietzsche.txt")
text = open(path).read()
chars = sorted(list(set(text)))
vocab_size = len(chars)+1
n_fac,cs,bs,n_hidden,n_fac = 42,8,64,256,42
char_indices = dict((c, i) for i, c in enumerate(chars))

idx = [char_indices[c] for c in text]

c_in_dat = [[idx[i+n] for i in range(0, len(idx)-1-cs, cs)]
            for n in range(cs)]
c_out_dat = [[idx[i+n] for i in range(1, len(idx)-cs, cs)]
            for n in range(cs)]

xs = [np.stack(c[:-2]) for c in c_in_dat] 
ys = [np.stack(c[:-2]) for c in c_out_dat]
x_rnn=np.stack(xs, axis=1)
y_rnn=np.expand_dims(np.stack(ys, axis=1), -1)

model=Sequential([
        Embedding(vocab_size, n_fac, input_length=cs, batch_input_shape=(bs,8)),
        BatchNormalization(),
        LSTM(n_hidden, return_sequences=True, stateful=True),
        TimeDistributed(Dense(vocab_size, activation='softmax')),
    ])
model.compile(loss='sparse_categorical_crossentropy', optimizer=Adam())
mx = len(x_rnn)//bs*bs
model.fit(x_rnn[:mx], y_rnn[:mx], batch_size=bs, epochs=4, shuffle=False)
```

Once I switched to the tensorflow backend it worked! But for the theano backend I still get nan's for the loss function.

My version of Keras and theano are:

```
In [2]: keras.__version__
Out[2]: '2.0.4'
In [4]: theano.__version__
Out[4]: '0.9.0'
```

**Comment by roebius** [comment]
Could you please do a last check: launch a Python console in a terminal window, enter _import theano_ and check the result. This way if there are any issues related to the GPU libraries several error messages should be displayed.

**Comment by dov** [comment]
Here is the output when using theano. Nothing apparently wrong. The rest of the deeplearning notebook seemed ok with theano.

```
$ python lstm_ex.py 
Using Theano backend.
WARNING (theano.sandbox.cuda): The cuda backend is deprecated and will be removed in the next release (v0.10).  Please switch to the gpuarray backend. You can get more information about how to switch at this URL:
 https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

Using gpu device 0: GeForce GTX 1070 (CNMeM is disabled, cuDNN 5105)
Epoch 1/4
75072/75072 [==============================] - 20s - loss: nan    
Epoch 2/4
75072/75072 [==============================] - 20s - loss: nan    
Epoch 3/4
75072/75072 [==============================] - 19s - loss: nan    
Epoch 4/4
75072/75072 [==============================] - 19s - loss: nan    
```
For comparison, here is the same when using the tensorflow backend:

```
$ python lstm_ex.py 
Using TensorFlow backend.
Epoch 1/4
2017-10-25 19:12:45.094644: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-10-25 19:12:45.094658: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-10-25 19:12:45.094664: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-10-25 19:12:45.094669: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-10-25 19:12:45.094673: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-10-25 19:12:45.192432: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-10-25 19:12:45.192716: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: GeForce GTX 1070
major: 6 minor: 1 memoryClockRate (GHz) 1.7845
pciBusID 0000:01:00.0
Total memory: 7.92GiB
Free memory: 7.71GiB
2017-10-25 19:12:45.192727: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-10-25 19:12:45.192731: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-10-25 19:12:45.192740: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1070, pci bus id: 0000:01:00.0)
 2112/75072 [..............................] - ETA: 28s - loss: 3.64082017-10-25 19:12:45.952428: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 20531 get requests, put_count=20505 evicted_count=1000 eviction_rate=0.0487686 and unsatisfied allocation rate=0.0548439
2017-10-25 19:12:45.952446: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
75072/75072 [==============================] - 9s - loss: 2.2584     
Epoch 2/4
75072/75072 [==============================] - 9s - loss: 1.9878     
Epoch 3/4
75072/75072 [==============================] - 9s - loss: 1.9102     
Epoch 4/4
75072/75072 [==============================] - 9s - loss: 1.8654     

```

**Comment by roebius** [comment]
This is weird, I was suspecting some gpu library issues but in this case everything seems to be ok, especially if you are able to successfully run the other portions of the notebook.
Out of curiosity, at this point I would also like to see what happens with a very small batch size.

Back to potential library issues: I tested the notebook with Theano using the gpuarray cuda backend, while in your case I see from the WARNING messages that Theano is using the old cuda backend. Have you tested also the new one? The installation is a bit tricky, but [in another issue at this link](https://github.com/roebius/deeplearning_keras2/issues/3) I provided an install file that you might adapt to suit your environment if you want to give it a try.

NOTE: I did not use Anaconda, but in case you did then the installation of gpuarray should be much simpler.
